File: Platform/Workspace/CSharp/Adapters/Allors.Workspace.Adapters.Remote/Workspace.cs
// <copyright file="Session.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace.Adapters.Remote
{
    using System;
    using System.Collections.Generic;
    using System.Net.Http;
    using Meta;

    public class Workspace : IWorkspace
    {
        private readonly Population population;

        private readonly Dictionary<long, long> workspaceIdByDatabaseId;

        private long worskpaceIdCounter;

        public Workspace(IMetaPopulation metaPopulation, Type instance, IWorkspaceStateLifecycle state, HttpClient httpClient)
        {
            this.MetaPopulation = metaPopulation;
            this.StateLifecycle = state;

            this.worskpaceIdCounter = 0;
            this.workspaceIdByDatabaseId = new Dictionary<long, long>();

            this.ObjectFactory = new ObjectFactory(this.MetaPopulation, instance);
            this.Database = new Database(this.MetaPopulation, httpClient);
            this.Sessions = new HashSet<Session>();

            this.population = new Population();
            this.WorkspaceClassByWorkspaceId = new Dictionary<long, IClass>();

            this.StateLifecycle.OnInit(this);
        }

        public IMetaPopulation MetaPopulation { get; }

        public IWorkspaceStateLifecycle StateLifecycle { get; }

        IObjectFactory IWorkspace.ObjectFactory => this.ObjectFactory;

        internal ObjectFactory ObjectFactory { get; }

        internal Database Database { get; }

        internal ISet<Session> Sessions { get; }

        public Dictionary<long, IClass> WorkspaceClassByWorkspaceId { get; }

        public ISession CreateSession() => new Session(this, this.StateLifecycle.CreateSessionState());

        internal void RegisterSession(Session session) => this.Sessions.Add(session);

        internal void UnregisterSession(Session session) => this.Sessions.Remove(session);

        internal void RegisterWorkspaceIdForDatabaseObject(long databaseId, long workspaceId) => this.workspaceIdByDatabaseId.Add(databaseId, workspaceId);

        internal void RegisterWorkspaceIdForWorkspaceObject(IClass @class, long workspaceId) => this.WorkspaceClassByWorkspaceId.Add(workspaceId, @class);

        internal long NextWorkspaceId() => --this.worskpaceIdCounter;
        
        internal long ToWorkspaceId(long id)
        {
            if (id > 0)
            {
                if (!this.workspaceIdByDatabaseId.TryGetValue(id, out var workspaceId))
                {
                    workspaceId = this.NextWorkspaceId();
                    this.workspaceIdByDatabaseId.Add(id, workspaceId);

                }

                return workspaceId;
            }

            return id;
        }

        internal void Set(Strategy strategy, IRoleType roleType, object value) => this.population.SetRole(strategy.WorkspaceId, roleType, value);

        internal object Get(Strategy strategy, IRoleType roleType)
        {
            this.population.GetRole(strategy.WorkspaceId, roleType, out var role);
            return role;
        }
    }
}

File: Core/Workspace/Typescript/libs/workspace/adapters/memory/src/lib/Workspace/Workspace.ts
import { IWorkspace, IWorkspaceLifecycle } from '@allors/workspace/domain/system';
import { Class, MetaPopulation, RelationType } from '@allors/workspace/meta/system';
import { Database } from '../Database/Database';
import { WorkspaceObject } from './WorkspaceObject';

export class Workspace implements IWorkspace {
  private readonly objectById: Map<number, WorkspaceObject>;

  constructor(public name: string, public metaPopulation: MetaPopulation, instance: {}, state: IWorkspaceLifecycle, httpClient: HttpClient) {
    this.objectFactory = new ObjectFactory(this.metaPopulation, instance);
    this.database = new Database(this.metaPopulation, httpClient, new Identities());

    this.workspaceClassByWorkspaceId = new Map();
    this.workspaceIdsByWorkspaceClass = new Map();

    this.objectById = new Map();

    this.lifecycle.OnInit(this);
  }

  objectFactory: ObjectFactory;

  createSession(): ISession {
    return new Session(this, this.state.createSessionContext());
  }

  database: Database;

  workspaceClassByWorkspaceId: Map<number, Class>;

  workspaceIdsByWorkspaceClass: Map<Class, number[]>;

  get(identity: number): WorkspaceObject {
    return this.objectById.get(identity);
  }

  registerWorkspaceObject(cls: IClass, workspaceId: number): void {
    this.workspaceClassByWorkspaceId.set(workspaceId, cls);

    let ids: Set<number> = this.workspaceIdsByWorkspaceClass.get(cls);
    if (!ids) {
      ids = [workspaceId];
    } else {
      ids = ids.add(ids, workspaceId);
    }

    this.workspaceIdsByWorkspaceClass[cls] = ids;
  }

  push(identity: number, cls: Class, version: number, changedRoleByRoleType: Map<RelationType, unknown>): void {
    let originalWorkspaceObject = this.objectById.get(identity);
    if (!originalWorkspaceObject) {
      this.objectById[identity] = new WorkspaceObject(this.database, identity, cls, ++version, changedRoleByRoleType);
    } else {
      this.objectById[identity] = new WorkspaceObject(originalWorkspaceObject, changedRoleByRoleType);
    }
  }
}
